### PR TITLE
VisualEditorPanel: finer grain listeners

### PR DESCRIPTION
**Pull request by xulman**
Hi,
I propose to add another layer of listeners into the `VisualEditorPanel`, and to rename to original ones (without breaking public API).

The original ones (renamed now into `itemChangedListeners`) were triggered everytime an item was changed in the GUI and were *not* triggered when the "Apply" button was pressed. The user-land code was therefore updated incrementally about *a presence of change* (no details about the change per se were provided via the listener notification **1**) and had to rescan somehow the GUI to find the change and update its own data. The user-land code never learnt that "apply" was pressed, so how should the code learn what has been changed and when to take these changes seriously?

**1**: Would be nice if the listeners would hear what triggered them -- what has been just changed. but we either break the API (by introducing param to the listeners) or create yet another type of listeners

The added listeners layer (`configChangedListeners`) is triggered only when "Apply" button is used, so they know that now is the right time to process the underlying `InputTriggerConfig` to readout the fresh new wanted state of key bindings...
(or I just totally misunderstood the whole thing......)

Cheers,
Vlado

### CODE REVIEW

**Comment by xulman** [comment]
The last commit f99fb9f12b22e4bc9a2d507b882f0c29533d0b7f was originally intended to become a PR on its own, but it is closely related to the former two commits.

The envisioned (example) usage is as follows, here in Kotlin:
```Kotlin
        //behaviourMap, config and (the associated) inputMap was defined earlier

        val cdb = CommandDescriptionBuilder()
        behaviourMap.keys().forEach { b -> cdb.addCommand(b,"all","") }
        //NB: don't know how to read out all behaviours' string IDs from config

        val editorPanel = VisualEditorPanel(config, cdb.get())

        val frame = JFrame(editorTitle)
        frame.contentPane.add(editorPanel)
        frame.pack()
        frame.isVisible = true

        editorPanel.addConfigCommittedListener { config.resetThisMapFor( inputMap, "all" ) }
        //with every "apply button" the config gets updated (as part of the functionality of the Editor)
        //with every "apply button", the associated inputMap gets updated too
```

File: src/main/java/org/scijava/ui/behaviour/io/InputTriggerConfig.java [hunk 1]
```diff
@@ -492,4 +492,41 @@ void addMap( final InputMap map, final String context )
...
+	{
+	    resetThisMap.clear();
+		for (String behaviourName : actionToInputsMap.keySet())
+		{
+			for (InputTrigger trigger : getInputs( behaviourName, context ))
+			{
+				resetThisMap.put( trigger, behaviourName );
+			}
+		}
+	}
```
Comment (tpietzsch) [inline src/main/java/org/scijava/ui/behaviour/io/InputTriggerConfig.java]: There is already functionality related to this.
First, this could be rewritten using an `InputTriggerAdder` which also takes care of corner cases (explicit `NOT_MAPPED` trigger, ...)
```java
	public void resetThisMapFor(final InputTriggerMap resetThisMap, final String context )
	{
		resetThisMap.clear();
		final InputTriggerAdder adder = inputTriggerAdder( resetThisMap, context );
		actionToInputsMap.keySet().forEach( adder::put );
	}
```

Second, there is `Behaviours.updateKeyConfig()` which provides very similar functionality.
https://github.com/scijava/ui-behaviour/blob/37d73656381bb698f455182b498b07fac5c4c459/src/main/java/org/scijava/ui/behaviour/util/Behaviours.java#L182-L223

You could *almost* rewrite `resetThisMapFor` as
```java
public void resetThisMapFor(final InputTriggerMap resetThisMap, final String context, final BehaviourMap behaviourMap )
{
	new Behaviours( resetThisMap, behaviourMap, null, context ).updateKeyConfig( this );
}
```
Note, the difference is the additional `behaviourMap` parameter. This will be used to decide which `behaviourName`s to retrieve from the `InputTriggerConfig`.

This is in contrast to your `resetThisMapFor`, which takes *all* `behaviourName`s that are defined in the `InputTriggerMap`. This will most likely work in the simple scenario where you just have a single `InputTriggerMap`/`BehaviourMap` pair, no chaining, blocking, etc.

You *could* replicate this as follows:
```java
public void resetThisMapFor(final InputTriggerMap resetThisMap, final String context)
{
	final BehaviourMap behaviourMap = new BehaviourMap();
	actionToInputsMap.forEach( ( action, inputs ) -> behaviourMap.put( action, new Behaviour() { /*dummy*/ } ) );
	new Behaviours( resetThisMap, behaviourMap, null, context ).updateKeyConfig( this );
}
```

However, I recommend that you use the actual `BehaviourMap` if you have it.

If you have a `Behaviours` instance which you use to add `Behaviour`s, then that manages an `InputTriggerMap`/`BehaviourMap` pair, already has a context and so on. In this case, I would try to keep this `Behaviours` instance around and simply do
```
behaviours.updateKeyConfig( theInputTriggerConfig );
```

In summary: Could you try to get the functionality you do via `createMapFor()` / `resetThisMapFor()` using the existing `updateKeyConfig` mechanism and report back? If that is feasible, I would not add `createMapFor()` and `resetThisMapFor()` to encourage `updateKeyConfig` as The Way&trade; to do updates.

**Comment by tpietzsch** [comment]
@xulman see inline comment about commit f99fb9f

The other changes regarding the VisualEditorPanel listeners are fine.

File: src/main/java/org/scijava/ui/behaviour/io/InputTriggerConfig.java (hunk 1 above)
Comment (xulman) [reply src/main/java/org/scijava/ui/behaviour/io/InputTriggerConfig.java]: hi, thanks for the detailed explanation, I have again learnt something new and I think I understand this package a bit better now, thank you indeed for your patience -- you create always super generic SW, that is sometimes difficult for ordinary programmers when they come to it with zero knowledge of it, it took me a while to see it as a whole... the behaviour map, input trigger map and config, and the "binder" InputTriggerAdder... and obviously I only thought I'm seeing it all, for example, I missed totally the existence of Behaviours... and most of the utils...

also IDE got me here... i was not reading docs, should have, my bad, I know, I relied instead on method names only, these can be sometimes tricky, need their context... here, it did not occur to me that `Behaviours.updateKeyConfig` actually updates Map and leaves the Config (its parameter) untouched

anyway,  please drop the commit of mine from this PR
I confirm, it was easy to use the updateKeyConfig mechanism: https://github.com/scenerygraphics/scenery/pull/324/commits/ed025a74a8189b1279c2646cc17592caa6de00d6
(except that it always creates a new object... that's failure on the scenery side, it could have been caching the aux objects next to the three maps)

**Comment by xulman** [comment]
please drop the commit https://github.com/scijava/ui-behaviour/pull/20/commits/f99fb9f12b22e4bc9a2d507b882f0c29533d0b7f of mine from this PR
I confirm, it was easy to use the updateKeyConfig mechanism: scenerygraphics/scenery@ed025a7

File: src/main/java/org/scijava/ui/behaviour/io/InputTriggerConfig.java (hunk 1 above)
Comment (tpietzsch) [reply src/main/java/org/scijava/ui/behaviour/io/InputTriggerConfig.java]: > I missed totally the existence of Behaviours... and most of the utils...

Yeah, this is not very discoverable. Basically the "preferred way of wiring up everything" has evolved over time (and is still evolving). With Mastodon it has crystallized that a good pattern is to maintain the `Behaviours` (and sibling `Actions`) as the "handles" for adding/updating input/action maps. (Older stuff , e.g., BDV, doesn't match that 100% yet but should be updated eventually.)

> also IDE got me here... i was not reading docs, should have, my bad, I know, I relied instead on method names only, these can be sometimes tricky, need their context... here, it did not occur to me that `Behaviours.updateKeyConfig` actually updates Map and leaves the Config (its parameter) untouched

Hmm, yes, that is very unfortunately named... :-( `Behaviours.updateWithKeyConfig(...)`, or maybe just `updateWith(...)` would be better. Do you have any other ideas?

**Comment by tpietzsch** [comment]
Ok, I dropped it. I also revised the listeners stuff to use `scijava-listeners`. So
```
editorPanel.addConfigChangeListener(listener)
```
becomes
```
editorPanel.configChangeListeners().add(listener)
```

And 
```
editorPanel.addConfigCommittedListener(listener)
```
becomes
```
editorPanel.configCommittedListeners().add(listener)
```

**Review by xulman — COMMENTED** [review]
@tpietzsch I'm worried we were way too fast here.... please, see https://github.com/scijava/ui-behaviour/issues/23
this is relevant to the commit https://github.com/scijava/ui-behaviour/pull/20/commits/8ebc51dd30f45d6da23b33cc15f7952ebe88c904

File: src/main/java/org/scijava/ui/behaviour/io/InputTriggerConfig.java (hunk 1 above)
Comment (xulman) [reply src/main/java/org/scijava/ui/behaviour/io/InputTriggerConfig.java]: aaaa, there's even `Actions`... I should really be paying more attention to "satellite" stuff around me, good know this. THANK YOU.

regarding the method name, what about `Behaviours.updateOwnMapFromThisConfig(...)` or perhaps without the pronoun "Own", or with? "updateIts" , "updateInternal", but the word Map probably would be worthwhile to have unless the method changes the whole object (like also the behaviours... )
